File: LibraryTests/C/test_factorial.c
#include <stdio.h>
#include <gmp.h>
#include <time.h>

// A very simple program to run factorial on large numbers with arbitrary precision
int main(int argc, char const *argv[])
{
    // Variables
    long startingValue, endingValue, step;
    mpz_t n, mult;
    clock_t start, end;
    double cpuTimeUsed;

    // Parameters
    printf("Insert the starting value: ");
    scanf("%ld", &startingValue);
    printf("Insert the ending value: ");
    scanf("%ld", &endingValue);
    printf("Insert the step: ");
    scanf("%ld", &step);

    // Initialize the file writer
    FILE* file = fopen("./CFactorialResults.csv","w");
    fprintf(file, "Value,Time Taken to Process Factorial (C)\n");
    fflush(file);

    //Calculating the specified number of times
    for (long trial = startingValue; trial <= endingValue; trial += step)
    {
        // Initialization
        mpz_init_set_ui(n, 1);
        mpz_init(mult);
        start = clock();
        // Actual Calculation
        for (long i = 1; i <= trial; i++)
        {
            mpz_set_ui(mult, i); // Have to initialize each time because mpz_mult requires mpz_t types
            mpz_mul(n, n, mult); // n = n * mult
        }
        end = clock();
        cpuTimeUsed = ((double)(end - start)) / CLOCKS_PER_SEC;
        long timeInMicroseconds = cpuTimeUsed * 1000000;

        // Printing the result
        printf("\nTime taken by %ld! = %ld microseconds\n", trial, timeInMicroseconds);
        fprintf(file,"%ld,%ld\n",trial, timeInMicroseconds);
        fflush(file);
    }

    fclose(file);
    printf("The test is over, check out the results in ./CFactorialResults\n");

    return 0;
}

File: LibraryTests/Java/TestFactorial.java
import java.io.File;
import java.io.PrintWriter;
import java.math.BigInteger;

class TestFactorial {
    public static void main(String[] args) {
        // Declaration and initialization
        Long startingValue = Long.parseLong(args[0]);
        Long upperBound = Long.parseLong(args[1]);
        Long step = Long.parseLong(args[2]);
        PrintWriter printWriter = null;

        try {
            printWriter = new PrintWriter(new File("./JavaFactorialResults.csv"));
            printWriter.write("Value,Time Taken to Process Factorial (Java)\n");
            printWriter.flush();
        } catch (Exception e) {
            e.printStackTrace();
            return;
        }

        for (Long trial = startingValue; trial <= upperBound; trial += step) {
            // Actual calculation
            BigInteger bigInteger = new BigInteger("1");
            long start = System.nanoTime();
            for (long i = 1; i <= trial; i++) {
                bigInteger = bigInteger.multiply(BigInteger.valueOf(i));
            }
            long end = System.nanoTime();

            // Write the result to the file
            printWriter.println(trial + "," + (end - start) / 1000);
            printWriter.flush();
            // Showing the result
            //System.out.println(trial + "! = " + bigInteger.toString());
            System.out.println("\nCalculation of " + trial + "! took " + (end - start) / 1000 + " microSeconds");

        }

        printWriter.close();
        System.out.println("Simulation Ended");

    }
}